### PR TITLE
display the output of -dtimings as a hierarchy

### PR DESCRIPTION
**Pull request by v-gb**
As briefly mentioned in https://github.com/ocaml/ocaml/pull/1027, this pull request changes the output of -dtimings from something like this:

```
all: 2.236s
parsing(hg.ml): 0.064s
-pp(hg.ml): 0.054s
-ppx(hg.ml): 0.000s
typing(hg.ml): 1.533s
transl(hg.ml): 0.220s
generate(sourcefile(hg.ml)): 0.418s
cmm(sourcefile(hg.ml)): 0.012s
compile_phrases(sourcefile(hg.ml)): 0.233s
selection(sourcefile(hg.ml)): 0.032s
comballoc(sourcefile(hg.ml)): 0.001s
cse(sourcefile(hg.ml)): 0.016s
deadcode(sourcefile(hg.ml)): 0.009s
spill(sourcefile(hg.ml)): 0.021s
split(sourcefile(hg.ml)): 0.013s
liveness(sourcefile(hg.ml)): 0.019s
regalloc(sourcefile(hg.ml)): 0.096s
linearize(sourcefile(hg.ml)): 0.000s
scheduling(sourcefile(hg.ml)): 0.000s
emit(sourcefile(hg.ml)): 0.019s
assemble(sourcefile(hg.ml)): 0.071s
```

to:

```
/: 2.236s
/parsing(hg.ml): 0.064s
/parsing/-pp(hg.ml): 0.054s
/parsing/-ppx(hg.ml): 0.000s
/typing(hg.ml): 1.533s
/transl(hg.ml): 0.220s
/generate(sourcefile(hg.ml)): 0.418s
/generate/cmm(sourcefile(hg.ml)): 0.012s
/generate/compile_phrases(sourcefile(hg.ml)): 0.233s
/generate/selection(sourcefile(hg.ml)): 0.032s
/generate/comballoc(sourcefile(hg.ml)): 0.001s
/generate/cse(sourcefile(hg.ml)): 0.016s
/generate/deadcode(sourcefile(hg.ml)): 0.009s
/generate/spill(sourcefile(hg.ml)): 0.021s
/generate/split(sourcefile(hg.ml)): 0.013s
/generate/liveness(sourcefile(hg.ml)): 0.019s
/generate/regalloc(sourcefile(hg.ml)): 0.096s
/generate/linearize(sourcefile(hg.ml)): 0.000s
/generate/scheduling(sourcefile(hg.ml)): 0.000s
/generate/emit(sourcefile(hg.ml)): 0.019s
/generate/assemble(sourcefile(hg.ml)): 0.071s
```

It makes parsing the output simpler, since one doesn't need to know that, say, cmm is part of generate.
The hierarchy in the output is reflected in the types, I don't know if that's really necessary. 

While I was looking at some outputs, I noticed that when linking, there were subcategories of /generate but not /generate itself, so I added /generate (I'll also point out the the call to the linker ends up in / right now, it may be worth splitting it into a separate category at some point).

### CODE REVIEW

**Comment by gasche** [comment]
Quick remarks:

- I think some spacing would help visualizing the hierarchy better. You could indent each entry by `2 * nesting_level` spaces (where `nesting_level` with your implementation is the length of the list); maybe you could just include the name of the child, and not the complete path, in this display strategy
- having the times *before* the names would ensure that times are aligned (if the Printf format is chosen to guarantee alignment), and thus easier to compare
- it would be nice if, in each category that is not a leaf, you could also display the "self time", the total time minus the sum of the children times.
- maybe instead of `All_*` you could use an option type, with `None` to represent the parent timer and `Some child` otherwise?
- If I understand correctly, the implementation currently does not guarantee that the conceptual nesting of timers corresponds to a nesting of timing calls at runtime. Would it be useful to check this?

**Comment by v-gb** [comment]
So with your first two points, we'd have something like this?

```
2.236s hg.ml
  0.064s parsing
    0.054s -pp
    0.000s -ppx
  1.533s typing
  0.220s transl
  0.418s generate
    0.012s cmm
    0.233s compile_phrases
    0.032s selection
    0.001s comballoc
    0.016s cse
    0.009s deadcode
    0.021s spill
    0.013s split
    0.019s liveness
    0.096s regalloc
    0.000s linearize
    0.000s scheduling
    0.019s emit
    0.071s assemble
```

I'm more interesting in parsing this output and combining many of them into a different format than reading it, but it sounds reasonable and probably not hard to parse either.

I am not sure about adding self time, because it makes some lines different. Having generated "other" subcategories would achieve the same result with less special cases when parsing though.

About these `All_x` constructors, I started with an option but it makes call sites a bit verbose for no real reason. And `Timings.Generate (None, ..)` is not particularly clear.

And yeah the nesting is not by construction. In fact I think compile_phrases should not be a leaf pass. Maybe it would make sense to pass around the prefix of the path, I haven't tried it.

**Comment by gasche** [comment]
Instead of `option` you could use

```ocaml
type 'a nesting = Self | Child of 'a
```

for callsite clarity.

(For the other suggestions, there are just suggestions; keep the ones that stick and feel free to ignore the others.)

**Comment by v-gb** [comment]
What about this version? Definitely add ?w=1 in the url of the diff, otherwise it's going to be unreadable.

The result looks like this:
```
$ time install/bin/ocamlopt -dtimings a.ml
0.218s a.ml
  0.010s parsing
    0.010s parser
  0.111s typing
  0.012s transl
  0.083s generate
    0.002s cmm
    0.025s compile_phrases
      0.001s selection
      0.002s comballoc
      0.001s cse
      0.001s liveness
      0.001s deadcode
      0.002s spill
      0.001s split
      0.003s regalloc
      0.002s emit
      0.011s other
    0.041s assemble
    0.015s other
  0.002s other
0.119s a.out
  0.001s cse
  0.001s deadcode
  0.001s liveness
  0.001s selection
  0.001s split
  0.005s spill
  0.017s regalloc
  0.018s assemble
  0.074s other
0.009s other
```

What I did is build the hierarchy dynamically, so what's printed definitely reflects what the code does.
It also avoids passing the current file around, which was sometimes passed around explicitely and sometimes mutable state.
It should make it easy to add subcategories, for instance for the time spent reading cmis.

I added this "other" row, which indicates that most of the link time is not detailed, and half of the time in compile_phrases is not detailed either (spent in Emit.data, presumably).
To avoid noise, I hide the 0.00s rows.

There's also an "other" at toplevel that I think mostly measures the cost of running the toplevel of the compiler. The "all" row is gone.

I didn't see the point of having all these variants mostly used in one place, so I replaced everything by strings.

**Comment by gasche** [comment]
I like the new output, but I'm not convinced by the use of strings. Each hardcoded string literal that occurs at least twice (for example `"transl"`) is an opportunity for silly typos that would have been caught, if you used variants, by the type-checker.

**Comment by v-gb** [comment]
We can have `val Timings.transl : string` for typos. It's not clear why we need a sum type of everything.

**Comment by gasche** [comment]
@chambart, as the author of the original code, would you have an opinion?

**Comment by v-gb** [comment]
I'll point out not having a sum type makes it easy to add one-off/temporary categories (this could also be done by having an `Other of string` case, though).

**Comment by xavierleroy** [comment]
@gasche , being well-read, is perhaps thinking of Perlis's epigram on programming #34:

> The string is a stark data structure and everywhere it is passed there is much duplication of process. It is a perfect vehicle for hiding information. 

(Perlis got the first Turing award.  Show some respect.)

In this particular case I agree that a string is fine, though.

**Comment by v-gb** [comment]
Alright, I'm done writing this pull request, so I squashed all the intermediate noise.

I shared constants like "generate", but they are still strings.
I tweaked a few things, like forcing the display of the toplevel "other" category, which is the same behavior as currently (where the "all" row is always shown).
I did a build of my repo with (a backport of) this pull request, and it looks sensible.
I ran the script below before and after to check the various ways of invoking the compiler all have sensible timings, and they have (relative to the master branch, because some of these calls like building the cmxa, or some of the bytecode ones don't show any details).
I now pad the durations with zeros instead of spaces, because otherwise there's no telling the difference between padding and indentation so the output is ambiguous.


```bash
set -x
# .mli
(set +x; for i in {0..1000}; do echo "val x : int Queue.t"; done > a.mli)
$d/ocamlc.opt -dtimings -c a.mli
$d/ocamlopt.opt -dtimings -c a.mli

# .ml
(set +x; for i in {0..1000}; do echo "let x = Queue.create ()"; done > a.ml)
$d/ocamlc.opt -dtimings -c a.ml
$d/ocamlopt.opt -dtimings -c a.ml

# a.cmxa
$d/ocamlc.opt -dtimings a.cmo -a -o a.cma
$d/ocamlopt.opt -dtimings a.cmx -a -o a.cmxa

# a.cmxa
$d/ocamlopt.opt -dtimings a.cmx -o a.cmxs

# a.out
$d/ocamlc.opt -dtimings a.cmo
$d/ocamlopt.opt -dtimings a.cmx

# pack
$d/ocamlc.opt -for-pack Foo -c a.mli
$d/ocamlc.opt -for-pack Foo -c a.ml
$d/ocamlc.opt a.cmo -pack -o foo.cmo -dtimings
$d/ocamlopt.opt -for-pack Foo -c a.mli
$d/ocamlopt.opt -for-pack Foo -c a.ml
$d/ocamlopt.opt a.cmx -pack -o foo.cmx -dtimings
```

**Comment by gasche** [comment]
@sliquister, could you rebase against trunk? Then I'd merge.

**Comment by v-gb** [comment]
I rebased, which dropped the initial commit updating .depend, squashed the commits, and moved the Changes entry from 4.05 to the next section.

**Comment by mshinwell** [comment]
@sliquister Sorry, please rebase again and then ping me, so we can get this in.

**Comment by v-gb** [comment]
@mshinwell rebased.

**Comment by gasche** [comment]
Merged, thanks!

**Comment by v-gb** [comment]
Thanks!

**Comment by gasche** [comment]
@sliquister: in case you are missing the warm embrace of the `-dtimings` codebase, I created an interesting [feature request](https://caml.inria.fr/mantis/view.php?id=7514) based on discussions with the Coq team. Just saying.
